### PR TITLE
Add BindMessageTemplate() and BindProperty() to ILogger

### PR DESCRIPTION
**Pull request by nblumhardt**
## Breaking change

Adds two methods to `ILogger`. Change 2 of 3.

## Details

A modification of Proposal 1 from #746

```csharp
MessageTemplate messageTemplate;
IEnumerable<LogEventProperty> properties;
if (!logger.BindMessageTemplate("Some {Number}", new object[] { 42 }))
    return;

var evt = logger.CreateEvent(DateTimeOffset.Now, level, exception, messageTemplate, properties);

LogEventProperty ticksProperty;
if (logger.CreateProperty("Ticks", 123, false, out ticksProperty))
    evt.AddOrUpdateProperty(ticksProperty);

LogEventProperty userProperty;
if (logger.CreateProperty("User", user, true, out userProperty))
    evt.AddOrUpdateProperty(userProperty);

logger.Write(evt);
```

Not the prettiest API in the world, but:

 * Keeps things orthogonal - there's already a public constructor on `LogEvent` so there's no need to return one
 * Maps 1:1 with what `Logger` internally needs to do, so adds little/no overhead
 * The methods can fail, but `ILogger` never throws - hence the `bool` return types

Notes:

Internally, `Logger.Write()` makes allowances for calls like:

```csharp
var myInts = new[] { 1, 2, 3 };
log.Information("Numbers are {Ints}", myInts);
```

https://github.com/serilog/serilog/blob/dev/src/Serilog/Core/Logger.cs#L174

The C# binder will consider the `int[]` as covariant with `object[]`, and so without our little hack, `Ints` ends up with the value `1`.

This behavior is _not_ propagated by `BindMessageTemplate()`, since that API is designed for programmatic scenarios we can't foresee and might break via the heuristic mentioned above.

### CODE REVIEW

File: src/Serilog/Core/Logger.cs [hunk 1]
```diff
@@ -168,21 +172,19 @@ public bool IsEnabled(LogEventLevel level)
...
-            var now = DateTimeOffset.Now;
-
             MessageTemplate parsedTemplate;
-            IEnumerable<LogEventProperty> properties;
-            _messageTemplateProcessor.Process(messageTemplate, propertyValues, out parsedTemplate, out properties);
+            IEnumerable<LogEventProperty> boundProperties;
+            _messageTemplateProcessor.Process(messageTemplate, propertyValues, out parsedTemplate, out boundProperties);
 
-            var logEvent = new LogEvent(now, level, exception, parsedTemplate, properties);
+            var logEvent = new LogEvent(DateTimeOffset.Now, level, exception, parsedTemplate, boundProperties);
```
Comment (merbla) [inline src/Serilog/Core/Logger.cs]: Should this be UTC?

Comment (merbla) [inline src/Serilog/Core/Logger.cs]: Ahh I see... previously was `DateTimeOffset.Now`

https://github.com/serilog/serilog/blob/dev/src/Serilog/Core/Logger.cs#L179

Comment (nblumhardt) [inline src/Serilog/Core/Logger.cs]: Yep; didn't see any reason for keeping the single-use variable :+1:

File: src/Serilog/Core/Logger.cs [hunk 2]
```diff
@@ -331,21 +335,19 @@ public void Write<T>(LogEventLevel level, Exception exception, string messageTem
         [MessageTemplateFormatMethod("messageTemplate")]
         public void Write(LogEventLevel level, Exception exception, string messageTemplate, params object[] propertyValues)
         {
-            if (messageTemplate == null) return;
             if (!IsEnabled(level)) return;
+            if (messageTemplate == null) return;
```
Comment (khellang) [inline src/Serilog/Core/Logger.cs]: 😄

Comment (nblumhardt) [inline src/Serilog/Core/Logger.cs]: Microoptimisation FTW ;-)

**Comment by nblumhardt** [comment]
Thanks!

**Pull request merged**